go/enclave/rollupchain/rollup_chain.go

```diff
@@ -909,6 +886,47 @@ func (rc *RollupChain) checkRollupsCorrectlyChained(rollup *core.Rollup, previou
 	return nil
 }
 
+// Checks the batch. If we've not seen a batch at this height before, we store it. If we have seen a batch at this
+// height before, we validate it against the other received batch at the same height.
+func (rc *RollupChain) checkAndStoreBatch(batch *core.Batch) error {
+	// We check the batch.
+	var txReceipts types.Receipts
+	// TODO - #718 - Determine what level of checking we should perform on the genesis batch.
+	if !batch.IsGenesis() {
+		var err error
+		txReceipts, err = rc.isInternallyValidBatch(batch)
+		if err != nil {
+			return fmt.Errorf("batch was invalid. Cause: %w", err)
+		}
+
+		// We check that we've stored the batch's parent.
+		if _, err = rc.storage.FetchBatch(batch.Header.ParentHash); err != nil {
+			return fmt.Errorf("could not retrieve parent batch. Cause: %w", err)
+		}
+	}
+
+	// If we've stored a batch at this height before, we ensure that it has the same transactions.
+	_, err := rc.storage.FetchBatchByHeight(batch.NumberU64())
+	if err != nil && !errors.Is(err, errutil.ErrNotFound) {
+		return fmt.Errorf("could not fetch batch. Cause: %w", err)
+	}
+	// TODO - #718 - Once the sequencer includes transactions deterministically (i.e. a batch of a given height always
+	//  contains the same transactions, regardless of reorgs), uncomment this check.
+	//if err == nil && batch.Header.TxHash != storedBatch.Header.TxHash {
+	//	return fmt.Errorf("two batches at same height did not have the same transactions")
+	//}
+
+	// We store the batch and update the head batch for that L1 block.
+	if err = rc.storage.StoreBatch(batch, txReceipts); err != nil {
+		return fmt.Errorf("failed to store batch. Cause: %w", err)
+	}
+	if err = rc.storage.UpdateHeadBatch(batch.Header.L1Proof, batch, txReceipts); err != nil {
```

Comment:
Wouldn't this rollback the head of the L2 chain if it was processing a historical rollup batch?